Thumbnail scrolling

will-moore opened the pull request:
See https://trello.com/c/JLpfTbmW/4-improve-thumbnails-loading-display

Instead of clicking up/down to load more thumbnails, the thumbnail panel is created at the full size (with only the visible thumbnails loaded) and more thumbnails are loaded when we scroll.
There are lots of different cases to handle and a lot has changed so everything needs testing:

 - Expected behaviour: any thumbnails that appear in the panel should be loaded (no empty grey squares).
 - While actually scrolling we don't load thumbnails, since that would load the whole dataset if you scroll fast from one end to the other. Instead, wait until scrolling slows / stops, then load thumbs for visible thumbnails.
 - You can see when loading is happening in the devtools Network tab.
 - In all scenarios below, check thumbs on load, then scroll about, then refresh thumbs (when scrollbar is in various locations).
 - Check that IE works.

**Scenarios to test:**

 - Single image in 1 or more Datasets, double-clicked in webclient. url like ```/webclient/img_detail/4413/?dataset=1301```
        - Check that thumbs loaded from *current* parent Dataset (not a different Dataset that the image is in)
        - Try double-clicking images in webclient near the start of the Dataset and at the end. Thumb panel in iviewer should scroll to the show the selected image.
        - Try with a small Dataset (< 5 images) and as large as possible.
        - Remove the ```?datasets=ID``` from the URL and reload. If image is in single Dataset, thumbs should be loaded. If more than 1 Dataset, thumbnail panel hidden.
 - Multiple selected images > Open with iviewer. Url like ```/iviewer/?images=4402,4403,4404,4405 ```
 - Orphaned image, double-clicked in webclient. e.g. ```webclient/img_detail/57012/``` should hide thumbnail panel.
 - Well - openwith > iviewer.  Url ```/iviewer/?well=148``` First image should be selected.
 - Datasets - openwith > iviewer. ```/iviewer/?dataset=10480``` First image selected.

mtbc commented:
I suspect that dropping IE11 (or not) should be a clear, conscious decision given that we often choose to support old stuff.

jburel commented:
Basic testing:
* Open dataset with < 10 Images by double-clicking on an image. The correct thumbnails were loaded
* Open a larger dataset (~80 images) visible thumbnails loaded. Scroll down to the bottom: more visible thumbnails are then loaded. Moving backward forward to load the rest of the thumbnails

jburel commented:
``
- Remove the ?datasets=ID from the URL and reload. If image is in single Dataset, thumbs should be loaded. If more than 1 Dataset, thumbnail panel hidden.
- Datasets - openwith > iviewer. /iviewer/?dataset=10480 First image selected.
- Well - openwith > iviewer. Url /iviewer/?well=148 First image should be selected.
- Orphaned image, double-clicked in webclient. e.g. webclient/img_detail/57012/ should hide thumbnail panel.
This works fine
``
Multiple selected images > Open with iviewer. Url like /iviewer/?images=4402,4403,4404,4405
``
If I select all the images in the dataset All the thumbnails are loaded
but If I open the dataset or double click on a thumbnail to open the viewer. Only few thumbnails are loaded.

jburel commented:
I don't think we can drop IE

jburel commented:
Safari:
* Clicking on well with only one field does not hide the thumbnails panel. It is empty
![pr-207_safari](https://user-images.githubusercontent.com/1022396/45495397-1a929d80-b76b-11e8-9861-6ce8b54a95b2.png)

jburel commented:
Noticeable loading difference between FF and chrome 
In FF I have time to see "Not Found"
even for small dataset

jburel commented:
FF 
Same problem as the one described in https://github.com/ome/omero-iviewer/pull/207#issuecomment-421030884

This time
it does not work for well with multiple fields.
![pr-2017-ff](https://user-images.githubusercontent.com/1022396/45496330-19fb0680-b76d-11e8-9f22-b4c4c785d40a.png)

mtbc commented:
I too saw that "Not Found": https://openmicroscopy.slack.com/archives/C0K5EED3R/p1535619171000100

jburel commented:
kymograph is correctly displayed cc @pwalczysko 
![pr-207-kymo](https://user-images.githubusercontent.com/1022396/45496908-5a0eb900-b76e-11e8-977a-d4f0fde1d61e.png)

jburel commented:
While testing 5.4.8 
Looking at data belonging to another user when that user has not viewed the image
leads to "Not found" in the left-hand panel and empty bird eye view
we should handle such case a bit better

will-moore commented:
@jburel Re: "If I select all the images in the dataset All the thumbnails are loaded
but If I open the dataset or double click on a thumbnail to open the viewer. Only few thumbnails are loaded."
As long as all the thumbnails that are *visible* are loaded then I think this is OK?
I assume that the number of images listed with ?image=1,2,3 is not going to be larger than a safe number of thumbnails to load (e.g. 200, or even more).

jburel commented:
All the thumbnails were loaded even the non visible when I do a selection of images
I probably select ~100

will-moore commented:
Also noticed that e.g. http://web-dev-merge.openmicroscopy.org/iviewer/?well=339085 (user-1) doesn't work (no thumbnails or image loaded).

will-moore commented:
Re: https://github.com/ome/omero-iviewer/pull/207#issuecomment-421319766 The 'Not Found' in this case is the same as previously (not related to this PR). To fix that we could use /webclient/render_thumbnail/ID instead of /webgateway/render_thumbnail. Then we'd get the 'No Image'  icon same as webclient?

will-moore commented:
Fixed the placeholder "Not Found" while images loading in Firefox.
Also fixed bugs with SPW images in a Well (```/webclient/img_detail/ID/```) and opening Well directly (```/iviewer/?well=ID```). In both cases, the thumbnails within the Well should be shown in the thumbnail_slider, even if it's only 1 field (same behaviour as before).

mtbc commented on src/app/thumbnail-slider.js:
```diff
@@ -700,6 +785,15 @@ export default class ThumbnailSlider extends EventSubscriber {
+            count = 0;
+        }
+        this.thumbnails.length = count;
+        this.thumbnails.fill({'version': 0, 'title': 'unloaded'});
```
Given https://github.com/ome/omero-iviewer/pull/207#issuecomment-421028262 is this okay?

jburel commented:
* Wells with 1 or multiple fields are now correctly displayed in FF,  Vivaldi and Safari
* I also select a large number of images, loading occurred when scrolling down the list

No cover yet: IE testing

jburel commented on src/app/thumbnail-slider.js:
```diff
@@ -100,25 +101,18 @@ export default class ThumbnailSlider extends EventSubscriber {
 
     /**
-     * the end index of the displayed thumbnails
+     * height of slider, so we which thumbs have scrolled into view
```
missing verb

jburel commented on src/app/thumbnail-slider.js:
```diff
@@ -141,22 +135,25 @@ export default class ThumbnailSlider extends EventSubscriber {
      * @param {BindingEngine} bindingEngine the BindingEngine (injected)
      */
-    constructor(context, element, bindingEngine) {
+    constructor(context, element, bindingEngine, taskQueue) {
```
Declare new parameter

jburel commented on src/app/thumbnail-slider.js:
```diff
@@ -314,94 +329,105 @@ export default class ThumbnailSlider extends EventSubscriber {
+    /**
+     * Loads unloaded thumbnails that scrolled into the thumbnail panel
+     *
+     * @param {number} scrollTop Current scroll position of the panel
```
declare parameter

jburel commented on src/app/thumbnail-slider.js:
```diff
@@ -480,35 +510,59 @@ export default class ThumbnailSlider extends EventSubscriber {
-     * Adds thumbnails to then be loaded.
+     * Adds thumbnails with Image IDs to this.thumbnails array.
+     * We create a new list from old, replacing the thumbnails (instead of
+     * modifying the old list) to make sure the changes is observed and
```
is/are

jburel commented on src/app/thumbnail-slider.js:
```diff
@@ -480,35 +510,59 @@ export default class ThumbnailSlider extends EventSubscriber {
+    /**
+     * Adds thumbnails with Image IDs to this.thumbnails array.
+     * We create a new list from old, replacing the thumbnails (instead of
+     * modifying the old list) to make sure the changes is observed and
```
is/are

jburel commented on src/app/thumbnail-slider.js:
```diff
@@ -480,35 +510,59 @@ export default class ThumbnailSlider extends EventSubscriber {
+     * modifying the old list) to make sure the changes is observed and
+     * UI updates.
+     *
+     * @param {Array.<Object>} thumbnails the thumbnails to be added
```
thumbnails not a parameter

jburel commented:
Testing in IE 11 
it works nicely too
